### DISCUSSION TITLE
How to change font face of the key's label?

### DISCUSSION THREAD

**Discussion by DivineDominion** [magit/transient#283] [Support] [answered]
I'm currently looping over all `transient-*` faces and decrease their size to 80% of the default font. That's usually fine for user interfaces with my large-ish default font setting.

I managed to get this far with a brute force change (screenshot from Magit's <kbd>?</kbd> transient):

![2024-04-05 13-51-28 Emacs - magit  dotfiles@2x](https://github.com/magit/transient/assets/59080/d4fcdafb-e6de-4745-a11c-e640a10630b4)

```elisp
(let ((transient-faces '(transient-active-infix transient-argument transient-delimiter transient-disabled-suffix transient-enabled-suffix transient-heading transient-higher-level transient-inactive-argument transient-inactive-value transient-inapt-suffix transient-key transient-key-exit transient-key-noop transient-key-return transient-key-stay transient-mismatched-key transient-nonstandard-key transient-unreachable transient-unreachable-key transient-value)))
    (dolist (face transient-faces)
      (set-face-attribute face nil :height 0.8)))
```

But as you can see the title/heading and keys shrink, the "Apply" etc. labels don't.

I couldn't figure out whether these are just unpropertized (aka use the default font) or not. Usually I go to the text I want to change and <kbd>C-u C-x =</kbd> and inspect the text properties that are in effect. Can't do that with a transient :) I also don't understand the source code tbh (Elisp novice)

Hints would be much appreciated!

**Comment by tarsius** [comment, marked as answer]
There is no dedicated face for this text. `default` is used.

Note that a lot of text in the transient buffer may be created using functions that are under the control of package authors. `transient` has no (okay limited) control over the faces used for such text. So even if we were to propertize all text, which is inserted using only basic `transient` features, there still would be other text that implicitly uses `default` or explicitly uses arbitary other faces.

So this has to be done my remapping the `default` face. There's already explicit support for doing that to control the pitch of text in this buffer.

Instead of adding yet another special case, I've added a new hook, `transient-setup-buffer-hook`, to allow doing this sort of thing. I've done that just now, so you'll have to update first. Then, you only need:

```elisp
(require 'face-remap)

(defun transient-i-shrunk-the-glyphs ()
  (face-remap-reset-base 'default) ;probably unnecessary
  (face-remap-add-relative 'default '((:height 100))))

(add-hook 'transient-setup-buffer-hook #'transient-i-shrunk-the-glyphs)
```

**Reply by tarsius** [reply]
6da27d7e1c16e85c3865e20ad5c81385eadc15dc

**Reply by DivineDominion** [reply]
Never replied, sorry! This works really well, thanks for the hook.

I had to shrink fixed-pitch, too, to get the key/argument fonts to shrink, too, but that could be a modus-themes thing.

```elisp
(defun transient-i-shrunk-the-glyphs ()
  (face-remap-reset-base 'default)
  (face-remap-add-relative 'default '((:height 120)))
  (face-remap-add-relative 'fixed-pitch '((:height 120))))
```
